stan/math/prim/prob/poisson_lcdf.hpp

```diff
@@ -52,7 +52,7 @@ return_type_t<T_rate> poisson_lcdf(const T_n& n, const T_rate& lambda) {
     return ops_partials.build(negative_infinity());
   }
 
-  const auto& log_Pi = log(gamma_q(n_val + 1.0, lambda_val));
+  const auto& log_Pi = to_ref(log(gamma_q(n_val + 1.0, lambda_val)));
```

Comment:
```suggestion
  const auto& log_Pi = to_ref_if<!is_constant_all<T_rate>::value
>(log(gamma_q(n_val + 1.0, lambda_val)));
```

This is actually a (performance) bug that needs fixing. All other changes you did are just forcing evaluation even when we don't need to evaluate and should be reverted. 

Anyway this should not trigger expression tests to fail, so I think changes from this PR will not help with that.

Reply:
@t4c1 agreed. I reverted this part. I think this had something to do with the rvalue `n_val + 1.0` (where n_val is an integer). I made the expression in the lambda in apply_scalar_binary capture by value instead of by reference.

Rok tells me you figured out something on this, so if it's something better happy to go with that. I just wanted to get my solution up so you'd be proud of me for figuring out rvalues and expressions.

Reply:
That something is actually almost exactly what you just pushed. I definitely am proud of you :D.

Reply:
However you should still include the `to_ref_if` I suggested here. While it wont affect tests passing it should make this a bit faster.

Reply:
:bowtie:

Oh I missed that that `to_ref_if` was separate. I added it.